src/main/java/org/ice4j/ice/ConnectivityCheckClient.java

```diff
@@ -944,6 +944,7 @@ protected void run()
                             "Pair failed: "
                                 + pairToCheck.toShortString());
                         pairToCheck.setStateFailed();
+                        updateCheckListAndTimerStates(pairToCheck);
```

Comment:
This runs inside the `synchronized (pairToCheck)` block, which seems like a bad idea (I can't prove it would cause a deadlock but it worries me).  Can you move it outside?